job-server/src/main/resources/reference.conf

```diff
@@ -0,0 +1,6 @@
+spark {
+  jobserver {
+    # Maximum number of dispatcher threads for each Spark context
+    max-dispatchers-per-context = 2
```

Comment:
Ekin,

I want to make sure there is no misunderstanding here.  The job server does
not simply allow running one "Spark job" in each Spark context, which
traditionally runs single-threaded.  Instead, each spark context, which is
mapped to a single JobManagerActor, may allow many "jobs" to be executed in
parallel, each one in their own thread.   For example, a user creates a
context in job server, which spins up a new JobManagerActor.  This context
contains RDDs to be queried.  Each query is a new Job Server job, executed
against this context.  Using the fair scheduler, we could easily have 16
threads running 16 separate Futures, each one containing a separate query
"job".

Job Server terminology: a "job" is not a single SparkContext, but rather a
single Future tied to a single thread.

As far as setJobGroup, it is the intention for each thread/Future to get
its own Jobgroup, because it maps to each JobServer job.

Hope this clears things up.  A thread pool of 2 or 1 thread would make this
important use case of the job server (multiple jobs within a context) not
work.

You can also see the above because there is a separate Atomic counter which
is meant to count the number of concurrent Futures allowed to run at the
same time, and it defaults to 4 I think.  Plus the design of NamedRdds,
which allows concurrent "jobs" within the same context to safely update
named shared RDDs.

This is the reason why originally we wanted to borrow the SparkContext's
ActorSystem threadpool, because we could execute multiple jobs within that
threadpool.

On Fri, Apr 3, 2015 at 3:35 AM, Ekin Gecikligun <notifications@github.com>
wrote:

> In job-server/src/main/resources/reference.conf
> <https://github.com/spark-jobserver/spark-jobserver/pull/111#discussion_r27723944>
> :
>
> > @@ -0,0 +1,6 @@
> > +spark {
> > +  jobserver {
> > +    # Maximum number of dispatcher threads for each Spark context
> > +    max-dispatchers-per-context = 2
>
> The reason for both is the same; Spark has its own threads for running
> tasks and we only use these threads to dispatch them to Spark for running.
> Hence the name "dispatcher" and low number of threads.
>
> A question we may ask is; since there are two threads but not a unique
> one, how will Spark react when we call setJobGroup. Then, if our threads
> are A, and B and the invocation is carried out by A; then only and all jobs
> dispatched by A would be given the specific id. We should perhaps reduce
> this to a single thread executor. What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/111/files#r27723944>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison